rules.py

```diff
@@ -583,6 +583,94 @@ def check(self, element):
                         element.sourceline, element.tag))
 
 
+class PartisanPrimary(base.BaseRule):
+    """Partisan elections should link to the correct political party. 
+    
+    For a NIST Election element of Election type primary, partisan-primary-open, 
+    or partisan-primary-closed, the Contests in that ContestCollection should 
+    have a PrimartyPartyIds that is present and non-empty.
+    """
+    election_type = None
+    
+    def __init__(self, election_tree, schema_file):
+        super(PartisanPrimary, self).__init__(election_tree, schema_file)
+        #There can only be one election element in a file
+        election_elem = self.election_tree.find("Election")
+        election_type_elem = election_elem.find("Type")
+        if election_type_elem is not None:
+            self.election_type = election_type_elem.text.strip()
+
+    def elements(self):
+        #only check contest elements if this is a partisan election
+        if self.election_type and self.election_type in (
+            "primary", "partisan-primary-open", "partisan-primary-closed"):
+            return ["CandidateContest"]
+        else:
+            return []
+
+    def check(self, element):
+        primary_party_ids = element.find("PrimaryPartyIds")
+        if (primary_party_ids is None or not primary_party_ids.text 
+                or not primary_party_ids.text.strip()):
+            raise base.ElectionError(
+                "Line %d. Election is of ElectionType %s but PrimaryPartyIds "
+                "is not present or is empty" % (
+                    primary_party_ids.sourceline, self.election_type))
+
+
+class PartisanPrimaryHeuristic(PartisanPrimary):
+    """Attempts to identify partisan primaries not marked up as such.
+    """
+    #add other strings that imply this is a primary contest
+    party_text = ["(DEM)", "(REP)"]
+
+    def elements(self):
+        if not self.election_type or self.election_type not in (
+            "primary", "partisan-primary-open", "partisan-primary-closed"):
+            return ["CandidateContest"]
+        else:
+            return []
+
+    def check(self, element):
+        contest_name = element.find("Name")
+        if contest_name is not None and contest_name.text is not None:
+            c_name = contest_name.text.replace(" ", "")
```

Comment:
To catch situations when there is a space between the parenthesis and party abbreviation e.g. "( Dem )"